packages/main/src/RPA/Email/Exchange.py

```diff
@@ -223,13 +255,37 @@ def authorize(
         kwargs["primary_smtp_address"] = (
             primary_smtp_address if primary_smtp_address else username
         )
-        self.credentials = Credentials(username, password)
+        if is_oauth:
+            self.credentials = OAuth2Creds(
+                on_token_refresh=self.on_token_refresh,
```

Comment:
I'm guessing there's some code I'm not following in the `oauthlib` library, but from what I see, we pass in a function that only logs to override the auto refresh method. Assuming I am wrong about this, you can ignore my comment and I'd appreciate learning how this is actually working.

Reply:
You're right 100% and at first glance this seems useless, but it actually adds support for customizing token refresh behaviour with ease in your robot, like in this [example](https://github.com/robocorp/example-oauth-email/pull/1/files#diff-b93aca6dec83011aab894b95b6deaaf19b48d5870212c8366f1e0e66acd3cc80R29-R33):
```python
class ExtendedExchange(Exchange):

    def __init__(self, *args, secret_name, **kwargs):
        super().__init__(*args, **kwargs)
        self._secret_name = secret_name

    def on_token_refresh(self, token):
        super().on_token_refresh(token)
        secret = vault.get_secret(self._secret_name)
        secret["token"] = dict(token)
        vault.set_secret(secret)
```
which saves the token in the vault on refresh so the new one can be picked up next time the robot decides to run again to authorize and send an e-mail.

Reply:
That's great... and now I'm thinking your comment should be added to the documentation as an advanced use case somehow! 😄

Reply:
Yes, not yet in the docs but exists in the docstring of the `on_token_refresh` method and most of the instructions will go in the Portal [bot](https://robocorp.com/portal/robot/robocorp/example-oauth-email) which show-cases the extension of `Exchange`. I need at least one customer dealing with OAuth2 flow on `RPA.Email.Exchange` through the library before committing any more focus into this. (as I bet most of them will still prefer or be able to use just "App Passwords")